cmd/kubeadm/app/cmd/phases/upgrade/apply/preflight.go

```diff
@@ -0,0 +1,102 @@
+	"k8s.io/kubernetes/cmd/kubeadm/app/features"
+	"k8s.io/kubernetes/cmd/kubeadm/app/phases/upgrade"
+	kubeadmutil "k8s.io/kubernetes/cmd/kubeadm/app/util"
+	configutil "k8s.io/kubernetes/cmd/kubeadm/app/util/config"
+)
+
+// RunPreflightChecksPhase runs all the preflight checks for upgrades
+func RunPreflightChecksPhase(data ApplyData) error {
+
+	client := data.Client()
+
+	err := upgrade.EnforceRequirements(client, data.Cfg(), data.IgnorePreflightErrors(), data.FeatureGates(), data.DryRun())
+	if err != nil {
+		return err
+	}
+
+	// Validate requested and validate actual version
+	klog.V(1).Infoln("[upgrade/apply] validating requested and actual version")
+	if err := configutil.NormalizeKubernetesVersion(&data.Cfg().ClusterConfiguration); err != nil {
+		return err
+	}
+
+	newK8sVersion, err := version.ParseSemantic(data.Cfg().KubernetesVersion)
+	if err != nil {
+		return errors.Errorf("unable to parse normalized version %q as a semantic version", data.Cfg().KubernetesVersion)
+	}
+
+	if err := features.ValidateVersion(features.InitFeatureGates, data.Cfg().FeatureGates, data.Cfg().KubernetesVersion); err != nil {
+		return err
+	}
+
+	// Enforce the version skew policies
+	klog.V(1).Infoln("[upgrade/version] enforcing version skew policies")
+	if err := EnforceUpgradeVersionPolicies(data.Cfg().KubernetesVersion, newK8sVersion, data.AllowExperimentalUpgrades(), data.AllowRCUpgrades(), data.Force(), data.VersionGetter()); err != nil {
+		return errors.Wrap(err, "[upgrade/version] FATAL")
+	}
+
+	// Use a prepuller implementation based on creating DaemonSets
+	// and block until all DaemonSets are ready; then we know for sure that all control plane images are cached locally
+	klog.V(1).Infoln("[upgrade/apply] creating prepuller")
+	prepuller := upgrade.NewDaemonSetPrepuller(client, data.Waiter(), &data.Cfg().ClusterConfiguration)
+	componentsToPrepull := constants.ControlPlaneComponents
+	if data.Cfg().Etcd.External == nil && data.UpgradeETCD() {
+		componentsToPrepull = append(componentsToPrepull, constants.Etcd)
+	}
+	if err := upgrade.PrepullImagesInParallel(prepuller, data.ImagePullTimeout(), componentsToPrepull); err != nil {
+		return errors.Wrap(err, "[upgrade/prepull] Failed prepulled the images for the control plane components error")
+	}
+
+	return nil
+}
+
+// EnforceUpgradeVersionPolicies makes sure that the version the user specified is valid to upgrade to
+// There are both fatal and skippable (with --force) errors
+func EnforceUpgradeVersionPolicies(newK8sVersionStr string, newK8sVersion *version.Version, allowExperimentalUpgrades bool, allowRCUpgrades bool, force bool, versionGetter upgrade.VersionGetter) error {
+	fmt.Printf("[upgrade/version] You have chosen to change the cluster version to %q\n", newK8sVersionStr)
+
+	versionSkewErrs := upgrade.EnforceVersionPolicies(versionGetter, newK8sVersionStr, newK8sVersion, allowExperimentalUpgrades, allowRCUpgrades)
+	if versionSkewErrs != nil {
+
```

Comment:
this empty line can be removed.

Reply:
I will as I think it's a bit odd after an `if` though I do have to say if formatting passes `gofmt` and `golint` I don't think we should complain about it in PR's but that is just my opinion.